src/filesystem/search_id.rs

```diff
@@ -0,0 +1,28 @@
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[cfg_attr(feature = "defmt-log", derive(defmt::Format))]
+/// Unique ID used to search for files and directories in the open File/Directory lists
+pub struct SearchId(pub(crate) u32);
+
+/// ID generator intented to be used in a static context.
+///
+/// This object will always return a different ID.
+pub struct IdGenerator {
+    next_id: core::sync::atomic::AtomicU32,
+}
+
+impl IdGenerator {
+    /// Create a new [`IdGenerator`].
+    pub const fn new() -> Self {
+        Self {
+            next_id: core::sync::atomic::AtomicU32::new(0),
+        }
+    }
+
+    /// Generate a new, unique [`SearchId`].
+    pub fn next(&self) -> SearchId {
+        use core::sync::atomic::Ordering;
+        let id = self.next_id.load(Ordering::Acquire);
```

Comment:
This isn't thread safe. You either need to use a critical-section (from the critical-section crate) or do fetch-add (which doesn't work on Cortex-M0 or RV32IMC).

Reply:
Also, unless you can prove otherwise, it's safest to use SeqCst ordering always. The consequences of getting it wrong are tricky subtle bugs, and the benefits of getting it right are maybe a slight performance bump that is irrelevant in a cold path of a microcontroller based filesystem.

Reply:
For thread safety, wouldn't it be more appropriate to have the ID generator as a member of the controller, rather than a static global? I believe the open files and (therefore the unique search ID) are specific to the controller only.

Reply:
You're right. I made `Controller` own an `IdGenerator` instead of a global static generator.